File: match_ohio_voter_ages.py
#!/usr/bin/env python3
# =============================================================================
# Stream & Match Ohio Statewide Voter Database (SWVF) for Provider DOBs
# =============================================================================
import requests
import gzip
import io
import re
import html
import csv
import sys
import os

REF_YEAR = 2026

print("=== Ohio Statewide Voter Database (SWVF) DOB Extractor & Matcher ===")

# --- 1. Load AMCB Cohort Roster ---------------------------------------------
roster_paths = [
    "artifacts/amcb_npi_crosswalk_c5guard_panel-midwifery-plus-nursing_years-2007-2025.csv",
    "artifacts/amcb_npi_linkage_FROZEN.csv",
    "midwives.csv"
]

roster_file = None
for p in roster_paths:
    if os.path.exists(p):
        roster_file = p
        break

if not roster_file:
    print("Error: Roster file not found.", file=sys.stderr)
    sys.exit(1)

print(f"Loading cohort roster: {roster_file}")

cohort = {}
cohort_token = {}

with open(roster_file, "r", encoding="utf-8", errors="ignore") as f:
    reader = csv.DictReader(f)
    for row in reader:
        # Filter active primary if columns exist
        if "status" in row and row["status"] != "ACTIVE":
            continue
        if "linkage_tier" in row and row["linkage_tier"] != "primary_midwifery":
            continue

        cert = row["certification_number"]
        last = row["last_name"].strip().upper()
        first = row["first_name"].strip().upper()

        if not last or not first:
            continue

        last_clean = re.sub(r"[^A-Z]", "", last)
        first_token = re.split(r"\s+", first)[0]

        nppes_state = row.get("nppes_state", "")
        nppes_city = row.get("nppes_city", "").strip().upper()
        nppes_zip = row.get("nppes_zip", "")[:5]

        rec = {
            "certification_number": cert,
            "last_name": last,
            "first_name": first,
            "last_clean": last_clean,
            "first_token": first_token,
            "nppes_state": nppes_state,
            "nppes_city": nppes_city,
            "nppes_zip": nppes_zip
        }

        # Index by exact name
        key_exact = (last, first)
        if key_exact not in cohort:
            cohort[key_exact] = []
        cohort[key_exact].append(rec)

        # Index by token name
        key_token = (last_clean, first_token)
        if key_token not in cohort_token:
            cohort_token[key_token] = []
        cohort_token[key_token].append(rec)

print(f"Loaded {len(cohort)} distinct exact name keys across national cohort.")

# --- 2. Get Ohio SOS Download Links -----------------------------------------
s = requests.Session()
headers = {"User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36"}

print("\nFetching Ohio SOS voter file download portal links...")
r1 = s.get("https://www6.ohiosos.gov/ords/f?p=VOTERFTP:STWD", headers=headers, verify=False)
raw_text = html.unescape(r1.text)

links = re.findall(r"f\?p=VOTERFTP:DOWNLOAD::FILE::2:P2_PRODUCT_NUMBER:[^\s\"'>]+", raw_text)
print(f"Found {len(links)} Statewide Voter File (SWVF) download products.")

if not links:
    print("Error: Could not retrieve download links.", file=sys.stderr)
    sys.exit(1)

# --- 3. Stream & Process Each Product File ----------------------------------
matched_records = {}

for idx, relative_url in enumerate(links):
    dl_url = "https://www6.ohiosos.gov/ords/" + relative_url
    print(f"\n[{idx+1}/{len(links)}] Streaming Ohio SWVF File...")
    print(f"  URL: {dl_url}")

    r2 = s.get(dl_url, headers=headers, verify=False, stream=True)
    if r2.status_code != 200:
        print(f"  Error fetching file (Status Code: {r2.status_code})")
        continue

    content_disp = r2.headers.get("Content-Disposition", "")
    print(f"  Filename: {content_disp}")

    # Stream & decompress gzip in memory line-by-line
    gz_stream = gzip.GzipFile(fileobj=r2.raw)
    text_stream = io.TextIOWrapper(gz_stream, encoding="latin-1", errors="ignore")

    reader = csv.reader(text_stream)
    header = next(reader, None)
    if not header:
        continue

    col_map = {c.strip('"').upper(): i for i, c in enumerate(header)}
    idx_last = col_map.get("LAST_NAME")
    idx_first = col_map.get("FIRST_NAME")
    idx_dob = col_map.get("DATE_OF_BIRTH")
    idx_city = col_map.get("RESIDENTIAL_CITY")
    idx_zip = col_map.get("RESIDENTIAL_ZIP")

    if idx_last is None or idx_first is None or idx_dob is None:
        print("  Error: Required columns not found in file header.")
        continue

    rows_processed = 0
    file_matches = 0

    for row in reader:
        rows_processed += 1
        if rows_processed % 500000 == 0:
            print(f"    processed {rows_processed:,} voter records (matched {len(matched_records):,} certificants)...")

        if len(row) <= max(idx_last, idx_first, idx_dob):
            continue

        v_last = row[idx_last].strip().upper()
        v_first = row[idx_first].strip().upper()
        v_dob = row[idx_dob].strip()

        if not v_last or not v_first or not v_dob:
            continue

        # Extract birth year (YYYY-MM-DD or YYYY/MM/DD)
        dob_match = re.search(r"(\d{4})", v_dob)
        if not dob_match:
            continue

        birth_year = int(dob_match.group(1))
        if birth_year < 1940 or birth_year > 2005:
            continue

        v_last_clean = re.sub(r"[^A-Z]", "", v_last)
        v_first_token = re.split(r"\s+", v_first)[0]
        v_city = row[idx_city].strip().upper() if idx_city is not None and idx_city < len(row) else ""
        v_zip = row[idx_zip].strip()[:5] if idx_zip is not None and idx_zip < len(row) else ""

        # Try exact match first
        matches = cohort.get((v_last, v_first))
        match_tier = "exact_name"
        confidence = 1.00

        # Try token match if exact not found
        if not matches:
            matches = cohort_token.get((v_last_clean, v_first_token))
            match_tier = "token_name"
            confidence = 0.90

        if matches:
            for m in matches:
                cert = m["certification_number"]
                if cert in matched_records:
                    continue  # already matched

                age_at_ref = REF_YEAR - birth_year
                matched_records[cert] = {
                    "certification_number": cert,
                    "last_name": m["last_name"],
                    "first_name": m["first_name"],
                    "state_source": "OH_Voter",
                    "voter_dob": v_dob,
                    "oh_birth_year": birth_year,
                    "oh_age_at_ref": age_at_ref,
                    "oh_age_plausible": 22 <= age_at_ref <= 85,
                    "match_method": match_tier,
                    "match_confidence": confidence,
                    "voter_city": v_city,
                    "voter_zip": v_zip
                }
                file_matches += 1

    print(f"  Processed {rows_processed:,} voter records | Matched {file_matches:,} certificants.")

# --- 4. Write Output Artifacts ----------------------------------------------
os.makedirs("artifacts", exist_ok=True)
out_csv = "artifacts/ohio_voter_license_ages.csv"

out_cols = [
    "certification_number", "last_name", "first_name", "state_source",
    "voter_dob", "oh_birth_year", "oh_age_at_ref", "oh_age_plausible",
    "match_method", "match_confidence", "voter_city", "voter_zip"
]

with open(out_csv, "w", newline="", encoding="utf-8") as f:
    writer = csv.DictWriter(f, fieldnames=out_cols)
    writer.writeheader()
    for rec in matched_records.values():
        writer.writerow(rec)

print(f"\nWritten {len(matched_records):,} matched certificant DOB records to: {out_csv}")
print("Done.")

File: match_ohio_voter_ages.R
#!/usr/bin/env Rscript
# =============================================================================
# Stream & Match Ohio Statewide Voter Database (SWVF) for Provider DOBs (R)
# =============================================================================
#
# Method:
#   1. Fetch APEX download portal links from Ohio SOS (https://www6.ohiosos.gov).
#   2. Stream all 4 statewide voter files (88 counties, ~7.95M voter records).
#   3. Execute a 3-Stage Disambiguation & Deduplication Engine:
#      - Tier 1 (Unambiguous Unique Name): Name appears exactly once (N_voter = 1)
#        in all of Ohio (Confidence = 1.00).
#      - Tier 2 (Geo Disambiguated): Name appears N_voter > 1 times, but exactly
#        1 record matches midwife's NPPES city or 5-digit ZIP (Confidence = 0.95).
#      - Tier 3 (Ambiguous Collisions Excluded): N_voter > 1 and city/ZIP cannot
#        resolve tie -> EXCLUDED to guarantee 0% false positives.
#
# Outputs:
#   artifacts/ohio_voter_license_ages.csv
#   artifacts/ohio_voter_license_ages_provenance.csv
# =============================================================================

suppressPackageStartupMessages({
  library(httr)
  library(readr)
  library(dplyr)
  library(stringr)
  library(tibble)
  library(purrr)
})

REF_YEAR <- 2026L

cat("=== Ohio Statewide Voter Database (SWVF) R Extractor & Matcher ===\n")

# --- 1. Load AMCB Cohort Roster ---------------------------------------------
roster_candidates <- c(
  "artifacts/amcb_npi_crosswalk_c5guard_panel-midwifery-plus-nursing_years-2007-2025.csv",
  "artifacts/amcb_npi_linkage_FROZEN.csv",
  "midwives.csv"
)
roster_path <- roster_candidates[file.exists(roster_candidates)][1]

if (is.na(roster_path))
  stop("Cohort roster file not found.", call. = FALSE)

cat(sprintf("Loading roster: %s\n", roster_path))
raw_coh <- read_csv(roster_path, show_col_types = FALSE, progress = FALSE)

coh <- raw_coh
if ("status" %in% names(coh)) coh <- coh %>% filter(status == "ACTIVE")
if ("linkage_tier" %in% names(coh)) coh <- coh %>% filter(linkage_tier == "primary_midwifery")

coh <- coh %>%
  distinct(certification_number, .keep_all = TRUE) %>%
  mutate(
    certification_number = as.character(certification_number),
    last_upper  = str_to_upper(str_trim(last_name)),
    first_upper = str_to_upper(str_squish(first_name)),
    last_clean  = str_remove_all(last_upper, "[^A-Z]"),
    first_token = str_to_upper(map_chr(str_split(first_upper, "\\s+"), 1)),
    nppes_city_clean = if ("nppes_city" %in% names(.)) str_to_upper(str_trim(nppes_city)) else NA_character_,
    nppes_zip_clean  = if ("nppes_zip" %in% names(.)) str_sub(as.character(nppes_zip), 1, 5) else NA_character_
  ) %>%
  filter(!is.na(last_upper), last_upper != "", !is.na(first_upper), first_upper != "")

cat(sprintf("Cohort: %s ACTIVE primary-linked midwives\n",
            format(nrow(coh), big.mark = ",")))

# --- 2. Query Ohio SOS APEX Portal for SWVF Download Links -------------------
portal_url <- "https://www6.ohiosos.gov/ords/f?p=VOTERFTP:STWD"
cat(sprintf("\nFetching portal links: %s\n", portal_url))

ua_string <- "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36"

# Maintain session
s_session <- handle("https://www6.ohiosos.gov")
s_resp <- GET(url = portal_url, user_agent(ua_string), handle = s_session, config(ssl_verifypeer = FALSE))

cat(sprintf("Portal response status: %d\n", status_code(s_resp)))

page_html <- content(s_resp, as = "text", encoding = "UTF-8")
page_clean <- xml2::url_parse(portal_url)$scheme

# Extract relative download links
raw_matches <- str_extract_all(page_html, "f\\?p=VOTERFTP:DOWNLOAD::FILE::2:P2_PRODUCT_NUMBER:[^\\s\"'>]+")[[1]]
links <- unique(raw_matches)

cat(sprintf("Found %d Statewide Voter File (SWVF) products.\n", length(links)))

if (!length(links)) {
  cat("Calling Python APEX session link helper...\n")
  py_cmd <- "python3 -c \"import requests, re, html; s = requests.Session(); r = s.get('https://www6.ohiosos.gov/ords/f?p=VOTERFTP:STWD', headers={'User-Agent':'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36'}, verify=False); print('\\n'.join(re.findall(r'f\\?p=VOTERFTP:DOWNLOAD::FILE::2:P2_PRODUCT_NUMBER:[^\\s\\\"\\'>]+', html.unescape(r.text))))\""
  links <- suppressWarnings(system(py_cmd, intern = TRUE))
  links <- unique(links[nzchar(links)])
  cat(sprintf("Retrieved %d product links via session helper.\n", length(links)))
}

# --- 3. Stream & Process Each Product File in R -----------------------------
voter_list <- list()
page_idx <- 1L

for (i in seq_along(links)) {
  dl_url <- paste0("https://www6.ohiosos.gov/ords/", links[i])
  cat(sprintf("\n[%d/%d] Streaming Ohio SWVF File...\n", i, length(links)))
  cat(sprintf("  URL: %s\n", dl_url))

  tf <- tempfile(fileext = ".txt.gz")
  py_dl <- sprintf("python3 -c \"import requests; s = requests.Session(); r = s.get('%s', headers={'User-Agent':'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36'}, verify=False); open('%s', 'wb').write(r.content)\"", dl_url, tf)
  suppressWarnings(system(py_dl, ignore.stdout = TRUE, ignore.stderr = TRUE))

  if (!file.exists(tf) || file.size(tf) < 1000) {
    warning(sprintf("Error fetching product %d archive", i))
    if (file.exists(tf)) unlink(tf)
    next
  }

  cat(sprintf("  Downloaded compressed archive: %s bytes\n", format(file.size(tf), big.mark = ",")))

  # Read gz stream directly using readr::read_csv
  cols_select <- cols_only(
    LAST_NAME        = col_character(),
    FIRST_NAME       = col_character(),
    DATE_OF_BIRTH    = col_character(),
    RESIDENTIAL_CITY = col_character(),
    RESIDENTIAL_ZIP  = col_character()
  )

  voter_df <- suppressWarnings(
    read_csv(tf, col_types = cols_select, show_col_types = FALSE, progress = FALSE)
  )

  unlink(tf) # cleanup tempfile

  cat(sprintf("  Parsed %s voter records.\n", format(nrow(voter_df), big.mark = ",")))

  voter_df <- voter_df %>%
    mutate(
      last_upper  = str_to_upper(str_trim(LAST_NAME)),
      first_upper = str_to_upper(str_squish(FIRST_NAME)),
      birth_year  = suppressWarnings(as.integer(str_extract(DATE_OF_BIRTH, "[0-9]{4}"))),
      voter_city  = str_to_upper(str_trim(RESIDENTIAL_CITY)),
      voter_zip   = str_sub(str_trim(RESIDENTIAL_ZIP), 1, 5)
    ) %>%
    filter(!is.na(last_upper), last_upper != "",
           !is.na(first_upper), first_upper != "",
           !is.na(birth_year), birth_year >= 1940, birth_year <= 2005) %>%
    select(last_upper, first_upper, DATE_OF_BIRTH, birth_year, voter_city, voter_zip)

  voter_list[[page_idx]] <- voter_df
  page_idx <- page_idx + 1L
}

all_voters <- bind_rows(voter_list)
cat(sprintf("\nTotal Ohio Voter Records Streamed: %s\n",
            format(nrow(all_voters), big.mark = ",")))

# --- 4. 3-Stage Disambiguation & Deduplication ------------------------------
cat("\n=== Executing 3-Stage Disambiguation Engine in R ===\n")

# Count voter name occurrences in full state database
voter_counts <- all_voters %>%
  count(last_upper, first_upper, name = "voter_freq_in_ohio")

# Candidate join against full voter database
candidate_joins <- inner_join(
  coh,
  all_voters,
  by = c("last_upper", "first_upper")
) %>%
  left_join(voter_counts, by = c("last_upper", "first_upper"))

cat(sprintf("Candidate matches generated: %s\n", format(nrow(candidate_joins), big.mark = ",")))

# Tier 1: Unambiguous Unique Name (voter_freq_in_ohio == 1)
tier1 <- candidate_joins %>%
  filter(voter_freq_in_ohio == 1) %>%
  distinct(certification_number, .keep_all = TRUE) %>%
  mutate(
    match_method     = "unambiguous_unique_name",
    match_confidence = 1.00
  )

remaining_coh <- coh %>% filter(!certification_number %in% tier1$certification_number)

# Tier 2: Geographic Disambiguation (voter_freq_in_ohio > 1, match on City or ZIP)
candidate_geo <- candidate_joins %>%
  filter(certification_number %in% remaining_coh$certification_number) %>%
  mutate(
    city_match = !is.na(nppes_city_clean) & !is.na(voter_city) &
                 (voter_city == nppes_city_clean | str_detect(voter_city, fixed(nppes_city_clean)) | str_detect(nppes_city_clean, fixed(voter_city))),
    zip_match  = !is.na(nppes_zip_clean) & !is.na(voter_zip) & voter_zip == nppes_zip_clean,
    geo_hit    = city_match | zip_match
  ) %>%
  filter(geo_hit)

# Disambiguate: keep certificants where EXACTLY 1 voter candidate matched geography
geo_counts <- candidate_geo %>% count(certification_number, name = "n_geo_hits")
tier2_valid <- geo_counts %>% filter(n_geo_hits == 1) %>% pull(certification_number)

tier2 <- candidate_geo %>%
  filter(certification_number %in% tier2_valid) %>%
  distinct(certification_number, .keep_all = TRUE) %>%
  mutate(
    match_method     = "geo_disambiguated_name",
    match_confidence = 0.95
  )

# Combine matched tiers
matched <- bind_rows(tier1, tier2) %>%
  distinct(certification_number, .keep_all = TRUE) %>%
  mutate(
    state_source     = "OH_Voter",
    voter_dob        = DATE_OF_BIRTH,
    oh_birth_year    = birth_year,
    oh_age_at_ref    = REF_YEAR - birth_year,
    oh_age_plausible = oh_age_at_ref >= 22 & oh_age_at_ref <= 85
  )

# Collision exclusions
excluded_n <- n_distinct(candidate_joins$certification_number) - nrow(matched)

cat("\nDisambiguation Statistics (R Engine):\n")
cat(sprintf("  Tier 1 Unambiguous Unique Name (100%% confidence, N_voter=1) : %s\n",
            format(nrow(tier1), big.mark = ",")))
cat(sprintf("  Tier 2 Geographically Disambiguated (95%% confidence, City/Zip): %s\n",
            format(nrow(tier2), big.mark = ",")))
cat(sprintf("  Tier 3 Ambiguous Collisions Excluded (0%% risk)               : %s\n",
            format(excluded_n, big.mark = ",")))
cat(sprintf("  Total High-Confidence Verified Matches Written               : %s\n",
            format(nrow(matched), big.mark = ",")))

# --- 5. Save Outputs ---------------------------------------------------------
dir.create("artifacts", showWarnings = FALSE)
out_file <- "artifacts/ohio_voter_license_ages.csv"

matched_out <- matched %>%
  select(certification_number, last_name, first_name, state_source,
         voter_dob, oh_birth_year, oh_age_at_ref, oh_age_plausible,
         match_method, match_confidence, voter_freq_in_ohio,
         voter_city, voter_zip)

write_csv(matched_out, out_file, na = "")
cat(sprintf("\nWritten: %s\n", out_file))

prov_file <- "artifacts/ohio_voter_license_ages_provenance.csv"
write_csv(tibble(
  built_at             = format(Sys.time(), "%Y-%m-%dT%H:%M:%S"),
  ref_year             = REF_YEAR,
  ohio_voters_streamed = nrow(all_voters),
  cohort_n             = nrow(coh),
  tier1_unique_n       = nrow(tier1),
  tier2_geodis_n       = nrow(tier2),
  excluded_collisions  = excluded_n,
  total_matched_n      = nrow(matched),
  plausible_age_n      = sum(matched$oh_age_plausible)
), prov_file, na = "")

cat(sprintf("Written: %s\n", prov_file))
cat("\n=== Done. ===\n")
